File: node/src/commands/issues/github-client.ts
/**
 * GitHub API client — wraps `gh` CLI for issue operations.
 *
 * Prefers `gh` CLI (handles auth, rate limits, pagination) over raw REST.
 * Falls back to GITHUB_TOKEN + REST if gh is unavailable.
 */
import { execFileSync } from "child_process";

export interface GitHubIssue {
  number: number;
  title: string;
  body: string;
  labels: string[];
  html_url: string;
  state: string;
}

export interface CreateIssueParams {
  repo: string;
  title: string;
  body: string;
  labels?: string[];
}

function ghAvailable(): boolean {
  try {
    execFileSync("gh", ["--version"], { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

export function createIssue(params: CreateIssueParams): GitHubIssue {
  const { repo, title, body, labels } = params;

  if (!ghAvailable()) {
    throw new Error(
      "GitHub CLI (gh) is required. Install: https://cli.github.com"
    );
  }

  const args = [
    "issue",
    "create",
    "--repo",
    repo,
    "--title",
    title,
    "--body",
    body,
  ];

  if (labels && labels.length > 0) {
    for (const label of labels) {
      args.push("--label", label);
    }
  }

  const result = execFileSync("gh", args, {
    encoding: "utf-8",
    stdio: ["ignore", "pipe", "pipe"],
  }).trim();

  // gh issue create returns the issue URL
  const urlMatch = result.match(
    /https:\/\/github\.com\/[^/]+\/[^/]+\/issues\/(\d+)/
  );
  const number = urlMatch ? parseInt(urlMatch[1], 10) : 0;

  return {
    number,
    title,
    body,
    labels: labels || [],
    html_url: result,
    state: "open",
  };
}

export function listOpenIssues(repo: string): GitHubIssue[] {
  if (!ghAvailable()) {
    return [];
  }

  try {
    const result = execFileSync(
      "gh",
      [
        "issue",
        "list",
        "--repo",
        repo,
        "--state",
        "open",
        "--json",
        "number,title,body,labels,url",
        "--limit",
        "200",
      ],
      { encoding: "utf-8", stdio: ["ignore", "pipe", "pipe"] }
    ).trim();

    if (!result) return [];

    const issues = JSON.parse(result);
    return issues.map((i: any) => ({
      number: i.number,
      title: i.title,
      body: i.body || "",
      labels: (i.labels || []).map((l: any) => l.name),
      html_url: i.url,
      state: "open",
    }));
  } catch {
    return [];
  }
}

File: python/rafter_cli/commands/issues/github_client.py
"""GitHub API client — wraps `gh` CLI for issue operations."""
from __future__ import annotations

import json
import shutil
import subprocess
from dataclasses import dataclass, field


@dataclass
class GitHubIssue:
    number: int
    title: str
    body: str
    labels: list[str] = field(default_factory=list)
    html_url: str = ""
    state: str = "open"


def _gh_available() -> bool:
    return shutil.which("gh") is not None


def create_issue(
    repo: str,
    title: str,
    body: str,
    labels: list[str] | None = None,
) -> GitHubIssue:
    """Create a GitHub issue via gh CLI."""
    if not _gh_available():
        raise RuntimeError(
            "GitHub CLI (gh) is required. Install: https://cli.github.com"
        )

    args = ["gh", "issue", "create", "--repo", repo, "--title", title, "--body", body]
    if labels:
        for label in labels:
            args.extend(["--label", label])

    result = subprocess.run(
        args, capture_output=True, text=True, check=True
    )
    url = result.stdout.strip()

    # Extract issue number from URL
    number = 0
    import re
    m = re.search(r"/issues/(\d+)", url)
    if m:
        number = int(m.group(1))

    return GitHubIssue(
        number=number,
        title=title,
        body=body,
        labels=labels or [],
        html_url=url,
        state="open",
    )


def list_open_issues(repo: str) -> list[GitHubIssue]:
    """List open issues in a repo via gh CLI."""
    if not _gh_available():
        return []

    try:
        result = subprocess.run(
            [
                "gh", "issue", "list",
                "--repo", repo,
                "--state", "open",
                "--json", "number,title,body,labels,url",
                "--limit", "200",
            ],
            capture_output=True, text=True, check=True,
        )
        if not result.stdout.strip():
            return []

        data = json.loads(result.stdout)
        return [
            GitHubIssue(
                number=i["number"],
                title=i["title"],
                body=i.get("body", ""),
                labels=[l["name"] for l in (i.get("labels") or [])],
                html_url=i.get("url", ""),
                state="open",
            )
            for i in data
        ]
    except (subprocess.CalledProcessError, json.JSONDecodeError):
        return []
